Raptor: Freq vj

antoine-de opened the pull request:
Split vehicle journeys in 2 different subclasses:
- DiscreteVehicleJourney
  
  the 'classic' vj
- FrequencyVehicleJourney
  
  the frequency vj with start, end and headway as additional informations

This has been done:
- to reduce the size of the VJ (we removed 3 useless fields for DiscreteVJ, the most used kind of VJ)
- to reduce the size of the departure/arrival raptor structures (we now only store discreteVJ stoptime, no frequencyVJ stoptime)
- to better handle frequencies (there was some bug in the way frequencies were handled)

The 2 kind of VJ are handled 2 different way in raptor (in best_stop_time to be precise):
- DiscreteVJ are handled as usual with a departure/arrival sorted vector on which we loop through to find the best stop time (seenext_valid_discrete_pick_up /previous_valid_discrete_drop_off)
- FrequencyVJ are handled for the moment in a naive way
  
  We loop through all the jpp's vj to find the next one.
  We can use more clever data structure if needed

antoine-de committed:
Work in progress

antoine-de committed:
Raptor: more tests are ok

antoine-de committed:
Routing: more tests are OK

antoine-de committed:
Type: start for serializing VJs

antoine-de committed:
Routing: switch back to ptr for VJ

There is too much pointer on the vjs, so we cannot handled them as
Value, we store a vector of unique_ptr in the journey_pattern

antoine-de committed:
Ed: add is_frequency to vj in db

antoine-de committed:
Routing: rebase on dev

antoine-de committed:
adapted: change for adapted

antoine-de committed:
Benchmark: add possibility to run again the same tests

kinnou02 commented on source/routing/best_stoptime.cpp:
```diff
@@ -46,79 +46,202 @@ best_stop_time(const type::JourneyPatternPoint* jpp,
     for(; idx < end; ++idx) {
         const type::StopTime* st = data.dataRaptor->st_forward[idx];
-        if (st->valid_end(reconstructing_path) && st->valid_hour(hour, true) &&
+        if (st->valid_end(reconstructing_path) &&
```
I prefer having the `&&` at the start of the line

antoine-de committed:
Routing: small changes

antoine-de committed:
Routing: change list to call to functor

change do not create a temporary list with all vjs, for we have a mean
to call a functor on all vjs.

TeXitoi commented on source/type/type.cpp:
```diff
@@ -379,6 +376,19 @@ void Calendar::build_validity_pattern(boost::gregorian::date_period production_p
+    if (this == &other) return false;
+    if(this->journey_pattern == other.journey_pattern){
+        // On compare les pointeurs pour avoir un ordre total (fonctionnellement osef du tri, mais techniquement c'est important)
+        return this->stop_time_list.front() < other.stop_time_list.front();
```
Je crois que le commentaire n'est plus d'actualité depuis que les stop_time sont des objects dans des vecteurs

TeXitoi commented on source/type/type.cpp:
```diff
@@ -552,6 +577,26 @@ std::vector<idx_t> VehicleJourney::get(Type_e type, const PT_Data &) const {
+JourneyPattern::~JourneyPattern() {}
+
+//we need to define the copy constructor because we don't want to copy the vjs
+//(idx can be spiped too)
```
I don't know the work spiped ;)

TeXitoi commented on source/routing/best_stoptime.cpp:
```diff
@@ -46,79 +46,194 @@ best_stop_time(const type::JourneyPatternPoint* jpp,
+                             bool reconstructing_path,
+                             const type::VehicleProperties &vehicle_properties,
+                             bool disruption_active) {
+    std::pair<const type::StopTime*, DateTime> best = {nullptr, DateTimeUtils::inf};
```
A previous  with a default to inf, that's strange, non?

TeXitoi commented on source/routing/best_stoptime.cpp:
```diff
@@ -126,53 +241,74 @@ earliest_stop_time(const type::JourneyPatternPoint* jpp,
+            next_valid_frequency_pick_up(jpp, dt, reconstructing_path, vehicle_properties, disruption_active);
+
+    // If no trip was found, we look for one the day after
+    if (first_discrete_st_pair.first || first_frequency_st_pair.first) {
```
That's quite complicated, if next_valid returns the good thing, just a "max" is enough.

TeXitoi commented on source/routing/best_stoptime.cpp:
```diff
@@ -126,53 +241,74 @@ earliest_stop_time(const type::JourneyPatternPoint* jpp,
-            throw navitia::exception("vj " + vj->uri + " has been ill constructed, it has no meta vj");
+    jpp->journey_pattern->for_each_vehicle_journey([&](const nt::VehicleJourney& vj ) {
+        if (! vj.meta_vj) {
+            throw navitia::exception("vj " + vj.uri + " has been ill constructed, it has no meta vj");
```
an assert is enough, no?

TeXitoi commented on source/routing/best_stoptime.cpp:
```diff
@@ -221,50 +358,143 @@ get_all_stop_times(const type::JourneyPatternPoint* jpp,
+    const int32_t x = std::ceil(diff / double(freq_vj.headway_secs));
+
+    std::cout << " next : " << str(lower_bound) << "," << str(upper_bound) << " diff= " << str(diff)
+              << " x = " << x << " res == " << DateTimeUtils::set(date, (lower_bound + x * freq_vj.headway_secs) % DateTimeUtils::SECONDS_PER_DAY) << std::endl;
```
trace to remove

antoine-de committed:
All: change after reviews

antoine-de committed:
Raptor: change after review
